File: typescript/src/ui/overlays/keycap.ts
import P5 from 'p5';
import Colors from '../colors';

const SIZE = 30;

export default class Keycap {
    private p5: P5;
    private position: P5.Vector;
    private angle: number;

    constructor(p5: P5, position: P5.Vector, angle: number = 0) {
        this.p5 = p5;

        this.position = position;
        this.angle = angle;
    }

    draw() {
        this.p5.push();

        this.p5.translate(this.position.x, this.position.y);

        this.p5.angleMode(this.p5.DEGREES);
        this.p5.rotate(this.angle);
        this.p5.angleMode(this.p5.RADIANS);

        this.p5.fill(Colors.BACKGROUND);
        this.p5.stroke(Colors.EDGE);
        this.p5.strokeWeight(1);

        this.p5.translate(-SIZE / 2, -SIZE / 2);

        this.p5.rect(0, 0, SIZE, SIZE, SIZE / 5);

        this.p5.fill(Colors.EDGE);
        this.p5.triangle(
            SIZE / 2,
            (SIZE / 10) * 3,
            (SIZE / 10) * 7,
            (SIZE / 10) * 7,
            (SIZE / 10) * 3,
            (SIZE / 10) * 7
        );

        this.p5.pop();
    }
}

File: rust/src/overlays/keycap.rs
use std::f64::consts::PI;

use web_sys::CanvasRenderingContext2d;

use crate::utils::{colors::Colors, config::Config, vector::Vector};

pub struct Keycap {
    position: Vector,
    angle: f64,
    config: Config,
}

impl Keycap {
    pub fn new(position: Vector, angle: f64) -> Keycap {
        let config = Config::new();

        Keycap {
            position,
            angle,
            config,
        }
    }
}

impl Keycap {
    pub fn draw(&self, canvas: CanvasRenderingContext2d) {
        canvas.save();

        let size = self.config.overlays.keycap.size;

        let half_size = size / 2.0;

        let _result = canvas.translate(self.position.x, self.position.y);
        let _result = canvas.rotate(PI / (180.0 / self.angle));

        canvas.stroke_rect(-half_size, -half_size, size, size);

        canvas.set_fill_style(&Colors::Edge.value().into());

        canvas.begin_path();
        canvas.move_to(0.0, -half_size + 8.0);
        canvas.line_to(half_size - 8.0, half_size - 8.0);
        canvas.line_to(-half_size + 8.0, half_size - 8.0);
        canvas.close_path();

        canvas.fill();

        canvas.restore();
    }
}
